beacon-chain/state/setters.go

```diff
@@ -0,0 +1,335 @@
+	b.merkleLayers[0][randaoMixes] = root[:]
+	b.recomputeRoot(int(randaoMixes))
+	b.lock.Unlock()
+	return nil
+}
+
+// SetSlashings --
+func (b *BeaconState) SetSlashings(val []uint64) error {
+	root, err := stateutil.SlashingsRoot(b.state.Slashings)
+	if err != nil {
+		return err
+	}
+	b.state.Slashings = val
+	b.lock.Lock()
+	b.merkleLayers[0][slashings] = root[:]
+	b.recomputeRoot(int(slashings))
+	b.lock.Unlock()
+	return nil
+}
+
+// SetPreviousEpochAttestations --
+func (b *BeaconState) SetPreviousEpochAttestations(val []*pbp2p.PendingAttestation) error {
+	root, err := stateutil.EpochAttestationsRoot(b.state.PreviousEpochAttestations)
+	if err != nil {
+		return err
+	}
+	b.state.PreviousEpochAttestations = val
+	b.lock.Lock()
+	b.merkleLayers[0][previousEpochAttestations] = root[:]
+	b.recomputeRoot(int(previousEpochAttestations))
+	b.lock.Unlock()
+	return nil
+}
+
+// SetCurrentEpochAttestations --
+func (b *BeaconState) SetCurrentEpochAttestations(val []*pbp2p.PendingAttestation) error {
+	root, err := stateutil.EpochAttestationsRoot(b.state.CurrentEpochAttestations)
+	if err != nil {
+		return err
+	}
+	b.state.CurrentEpochAttestations = val
+	b.lock.Lock()
+	b.merkleLayers[0][currentEpochAttestations] = root[:]
+	b.recomputeRoot(int(currentEpochAttestations))
+	b.lock.Unlock()
+	return nil
+}
+
+// SetJustificationBits --
+func (b *BeaconState) SetJustificationBits(val bitfield.Bitvector4) error {
+	root := bytesutil.ToBytes32(b.state.JustificationBits)
+	b.state.JustificationBits = val
+	b.lock.Lock()
+	b.merkleLayers[0][justificationBits] = root[:]
+	b.recomputeRoot(int(justificationBits))
+	b.lock.Unlock()
+	return nil
+}
+
+// SetPreviousJustifiedCheckpoint --
+func (b *BeaconState) SetPreviousJustifiedCheckpoint(val *ethpb.Checkpoint) error {
+	root, err := stateutil.CheckpointRoot(b.state.PreviousJustifiedCheckpoint)
+	if err != nil {
+		return err
+	}
+	b.state.PreviousJustifiedCheckpoint = val
+	b.lock.Lock()
+	b.merkleLayers[0][previousJustifiedCheckpoint] = root[:]
+	b.recomputeRoot(int(previousJustifiedCheckpoint))
+	b.lock.Unlock()
+	return nil
+}
+
+// SetCurrentJustifiedCheckpoint --
+func (b *BeaconState) SetCurrentJustifiedCheckpoint(val *ethpb.Checkpoint) error {
+	root, err := stateutil.CheckpointRoot(b.state.CurrentJustifiedCheckpoint)
+	if err != nil {
+		return err
+	}
+	b.state.CurrentJustifiedCheckpoint = val
+	b.lock.Lock()
+	b.merkleLayers[0][currentJustifiedCheckpoint] = root[:]
+	b.recomputeRoot(int(currentJustifiedCheckpoint))
+	b.lock.Unlock()
+	return nil
+}
+
+// SetFinalizedCheckpoint --
+func (b *BeaconState) SetFinalizedCheckpoint(val *ethpb.Checkpoint) error {
+	root, err := stateutil.CheckpointRoot(b.state.FinalizedCheckpoint)
+	if err != nil {
+		return err
+	}
+	b.state.FinalizedCheckpoint = val
+	b.lock.Lock()
+	b.merkleLayers[0][finalizedCheckpoint] = root[:]
+	b.recomputeRoot(int(finalizedCheckpoint))
+	b.lock.Unlock()
+	return nil
+}
+
+func (b *BeaconState) recomputeRoot(idx int) {
```

Comment:
Please add a comment that says caller must hold the lock before calling this method. Since we can't enforce that the lock is in use, we can at least warn the user to use this appropriately :man_shrugging: